[WIP] Allow deploying pacemaker on SLE12

vuntz opened the pull request.

vuntz committed:
Stop restricting pacemaker roles on SUSE to SLE11

vuntz committed:
corosync: Update packages for SLE12

There's no openais package anymore, and the release package is
sle-ha-release.

vuntz committed:
pacemaker: Disable hb_gui code on SLE12

It's gone there.

vuntz committed:
corosync: Update corosync.conf template for corosync 2

Main change is the node list in udpu, which moved in its own section.

vuntz committed:
pacemaker: Start pacemaker service on SLE12

This is new in SLE12.

Also, we need to do that before checking if cluster is online, as
corosync just quits if nothing requiring runs.

vuntz committed:
pacemaker: Only enable pacemaker service when needed

With the require_clean_for_autostart logic, we do not want the service
to be started on boot as it would start corosync, and bypass the logic
where chef finds out whether it's okay to start it or not.

vuntz committed:
corosync: Do not start corosync on SLE12

The corosync.service unit has StopWhenUnneeded, which really means that
it should not be started standalone. Instead, we start the pacemaker
service now, which will automatically start corosync (as it's marked as
a dependency).

vuntz committed:
corosync: Use a systemd service file for corosync-shutdown-cleaner

tboerger commented on chef/cookbooks/corosync/attributes/default.rb:
```diff
@@ -23,13 +23,17 @@
 when 'suse'
-  default[:corosync][:platform][:packages] = %w(sle-hae-release corosync openais)
+  if node.platform_version.to_f >= 12.0
+    default[:corosync][:platform][:packages] = %w(sle-ha-release corosync)
```
No openais required for SLE 12?

vuntz replied:
Right, it's gone.

tboerger commented on chef/cookbooks/corosync/recipes/service.rb:
```diff
@@ -111,23 +111,42 @@
-    action :nothing
-    subscribes :run, resources(:template=> "/etc/init.d/#{corosync_shutdown}"), :delayed
+    # Make sure that any dependency change is taken into account
+    bash "insserv #{corosync_shutdown} service" do
```
Why have we done that with ```bash``` instead of the ```service``` resource?

vuntz replied:
It's not about enabling the service, but about making sure the dependency chain is taken into account if there's any dependency change. (Btw, that's not new code, just indentation change here)

tboerger replied:
I know that this is not new code :)

tboerger commented on chef/cookbooks/corosync/templates/default/corosync.conf.v2.erb:
```diff
@@ -0,0 +1,107 @@
+totem {
+	version: 2
+
+        cluster_name: <%= @cluster_name %>
```
Different indention

vuntz replied:
Fixed (and other indentation bits too).

tboerger commented on chef/cookbooks/corosync/templates/default/corosync.conf.v2.erb:
```diff
@@ -0,0 +1,107 @@
+		# addresses across multiple Corosync clusters sharing
+		# the same network.
+<% if @transport == 'udp' -%>
+                mcastaddr: <%= @mcast_addr %>
```
Indentions

tboerger commented on chef/cookbooks/corosync/templates/default/corosync.conf.v2.erb:
```diff
@@ -0,0 +1,107 @@
+	# over. If you define more than one interface stanza, you must
+	# also set rrp_mode.
+	interface {
+                # Rings must be consecutively numbered, starting at 0.
```
Indention

tboerger commented on chef/cookbooks/corosync/templates/default/corosync.conf.v2.erb:
```diff
@@ -0,0 +1,107 @@
+		# messaging, and also the immediately preceding
+		# port. Thus if you set this to 5405, Corosync sends
+		# messages over UDP ports 5405 and 5404.
+                mcastport: <%= @mcast_port %>
```
Indention

tboerger commented on chef/cookbooks/corosync/templates/default/corosync.conf.v2.erb:
```diff
@@ -0,0 +1,107 @@
+<% if @transport == 'udpu' -%>
+nodelist {
+  <% @members.sort.each do |memberaddr| -%>
+    node {
```
Indention

tboerger commented on crowbar_framework/app/models/pacemaker_service.rb:
```diff
@@ -33,14 +33,14 @@ def role_constraints
           "count" => 32,
           "platform" => {
-            "suse" => "/11\.[3-9]/"
+            "suse" => "/.*/"
```
You enable it only for SUSE, but within the cookbooks you've got checks if it's not SUSE?

vuntz replied:
Yes, the cookbooks are not fully ready for non-SUSE (and this was actually never tested), so we make it impossible to use that in the webui for now.

aspiers commented on chef/cookbooks/pacemaker/recipes/default.rb:
```diff
@@ -26,18 +26,20 @@
-  node[:pacemaker][:platform][:graphical_packages].each do |pkg|
-    package pkg
-  end
+unless node.platform == 'suse' && node.platform_version.to_f >= 12.0
```
This seems a bit convaluted, I am not even sure whether hb_gui is available outside SUSE.

aspiers commented on chef/cookbooks/pacemaker/recipes/default.rb:
```diff
@@ -53,6 +55,24 @@
   include_recipe "corosync::default"
 end
 
+if (platform_family?("suse") && node.platform_version.to_f >= 12.0) || platform_family?("rhel")
```
This is getting kind of ugly. Maybe it would be better to break this into smaller distro-specific recipes, then we can just ignore the rhel one.

vuntz committed:
Add validation check that all nodes in cluster have same platform

It's highly unlikely that nodes using two different platforms can really
be part of the same cluster, so disallow that.

vuntz committed:
crowbar-pacemaker: boot.lvm service is SLE11 only

vuntz committed:
crowbar-pacemaker: Update kernel module config for SLE12

The way to load modules on boot changed with systemd.

vuntz committed:
drbd: Only start the service if some resources are defined

In SLE12, the service fails to start if no resources are defined.

vuntz committed:
crowbar-pacemaker: Only enable drbd after we've configured resources

In SLE12, drbd fails to start if there are no resources. So we want to
make sure that the resources are defined first. Of course, to configure
resources, we still need the packages to be installed and the global
config, so we now include these drbd recipes (which we were using
indirectly through drbd::default before).

Note that the start itself actually has no real effect as there's a
restart when we configure the resources; however, we still want to
enable the drbd service.

vuntz committed:
pacemaker: Fix parsing of resources with no meta

Newer versions of crmsh now always include lines such as meta even if
there's nothing in there. Our parser was not expecting this.

vuntz committed:
crowbar-pacemaker: Fix RA for haproxy for SLE12

There's no LSB init script anymore, but a systemd service file.

vuntz commented:
This is slowly getting to a state where things are working. Current set of issues:
* crmsh now hides the `ocf:heartbeat:` prefix from RAs that have it; two possible solutions: use `crm configure show xml $resource` to parse the output, or change our cookbooks to stop using that prefix. The former is cleaner but requires more effort.
* not really happy with the current `sle-ha-release` package we install, as it comes with tons of recommends that makes the install very slow. We could stop installing it, or install it without recommends (which means not using the `package` resource), or install it before any sync mark so that slowness doesn't impact us.
* when using shared storage with NFS, we hit https://www.novell.com/support/kb/doc.php?id=7014266 because SLE12 uses NFSv4 by default now -- technically, it's not our fault, but I wonder if we could do something to help.

tboerger commented:
> crmsh now hides the ocf:heartbeat: prefix from RAs that have it; two possible solutions: use crm configure show xml $resource to parse the output, or change our cookbooks to stop using that prefix. The former is cleaner but requires more effort.

HAWK is built around the XML parsing :)

tboerger commented:
> not really happy with the current sle-ha-release package we install, as it comes with tons of recommends that makes the install very slow. We could stop installing it, or install it without recommends (which means not using the package resource), or install it before any sync mark so that slowness doesn't impact us.

Or we should extend the package resource to append something to the install command of the package resource. If i remember correctly this is possible for the gem_package resource.

mjura commented:
It was tested by me with OpenStack HA instances feature. I didn't have any problems.
Works fine +1

aspiers commented:
This PR does have issues, and I am reworking it as part of https://trello.com/c/ql5ui9s2

Currently you can see my branch at https://github.com/crowbar/barclamp-pacemaker/compare/master...aspiers:sle12

tboerger commented on chef/cookbooks/crowbar-pacemaker/templates/default/crowbar-drbd.conf.erb:
```diff
@@ -0,0 +1,4 @@
+# load modules for drbd at boot time
+<% @modules.each do |mod| -%>
+<%= @mod %>
```
Don't forget to fix that, the @ in front of mod is wrong.
